C-API cleanup: moved cvErrorStr to new interface, minor ts changes

mshabunin opened the pull request:
**Note:** `toString` might be too generic name (even though it is in `cv::Error::` namespace), another variant is `codeToString` (we have `typeToString` and `depthToString` in check.hpp).

**Note:** _ts_ module seem to have no other C API usage except for `ArrayTest` class which requires refactoring.

asmorkalov commented:
What about ErrorMessage()?

mshabunin commented:
> What about ErrorMessage()?

We usually start functions with lowercase, so it would be `errorMessage`. Should I leave it in the `cv::Error` namespace then? `cv::Error::errorMessage(...)` or `cv::errorMessage(...)`.

asmorkalov commented:
No objections, up to you.

opencv-alalek commented on modules/core/include/opencv2/core/base.hpp:
```diff
@@ -122,6 +122,10 @@ enum Code {
 };
+
+/** Retrieves textual description of the error given its code */
+CV_EXPORTS const char* toString(int status);
```
`int` -> `Code` at least to enable C++ ADL without explicit namespace.

mshabunin commented on modules/ts/src/ts.cpp:
```diff
@@ -544,7 +542,7 @@ static int tsErrorCallback( int status, const char* func_name, const char* err_m
     TS* ts = (TS*)data;
     const char* delim = std::string(err_msg).find('\n') == std::string::npos ? "" : "\n";
-    ts->printf(TS::LOG, "OpenCV Error:\n\t%s (%s%s) in %s, file %s, line %d\n", cvErrorStr(status), delim, err_msg, func_name[0] != 0 ? func_name : "unknown function", file_name, line);
+    ts->printf(TS::LOG, "OpenCV Error:\n\t%d (%s%s) in %s, file %s, line %d\n", status, delim, err_msg, func_name[0] != 0 ? func_name : "unknown function", file_name, line);
```
Note: used numeric code here instead of string

mshabunin commented:
I've changed interface to use enum instead of int and it lead to some extra changes (+_opencv_contrib_). Also extracted two new headers in _core_ module: exception.hpp and fwddecl.hpp.

asmorkalov reviewed (approved):
:+1:

mshabunin commented:
Sorry, I updated my branch but forgot to create new PR in opencv_contrib. Here it is https://github.com/opencv/opencv_contrib/pull/3789

mshabunin commented:
OH, I just found your PR have been merged with the same fix. Sorry again :face_with_diagonal_mouth:
